feat: add OTEL_LOG_LEVEL env var

naseemkullah opened the pull request:
User can now control the log level by means of env var.

Addresses JS implementation of  https://github.com/open-telemetry/opentelemetry-specification/issues/569

naseemkullah commented:
How can we get the test passing for `node-12: browsers`? Getting: `ReferenceError: process is not defined`.

naseemkullah commented:
Changed default level based on changes in https://github.com/open-telemetry/opentelemetry-js/pull/856/

Is there a way to avoid specifying the default log level in tracing and metrics packages and instead just inherit it from the the core package?

vmarchaud commented:
@naseemkullah `process` is only available within node, you need to have a different mecanism for the browser. We already have multiple functions that have different implementation depending on the platform [here](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-core/src/platform/)

naseemkullah commented:
CI is not triggering 🤔

naseemkullah commented:
> @naseemkullah `process` is only available within node, you need to have a different mecanism for the browser. We already have multiple functions that have different implementation depending on the platform [here](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-core/src/platform/)

Thanks @vmarchaud, I've refactored accordingly.

dyladan commented on packages/opentelemetry-core/src/platform/browser/log-level.ts:
```diff
@@ -0,0 +1,24 @@
+ */
+import { LogLevel } from '../../common/types';
+
+export function getLogLevel(): LogLevel {
```
Log level is always info for browser? @obecny what is the typical way to configure things like this

obecny replied:
By default info, warning and error are always shown so this will be fine and we will have the same behaviour for node and browser. 
I'm wondering a different thing - if we could implement the same mechanism and check global "OTEL_LOG_LEVEL" on `window` so mimic the same behaviour. But this might be something we can discuss on Wednesday perhaps. If we allow "globals" to have impact on otel then maybe we should have the same using window. There might be some cons with this but this is just an idea and not in this PR. Maybe @draffensperger can add some input here on the idea to mimic env variables as globals on `window` ?

draffensperger replied:
Do we still allow the parent trace ID and span ID to be set via a window variable from the server? If so, there is some precedent of allowing configuration via a window var.

The nice thing about making it available to be set on the window is that it allows a workflow where we build the OTel library and someone gets it as a `<script>` tag via a CDN and then adds configuration to it via a separate script tag, sort of like this https://github.com/census-instrumentation/opencensus-web#using-via-a-script-tag where I set the sampling rate based on window vars.

draffensperger replied:
I think if we do that, we should have some prefix for OTel config settings for the variables on the window like `_otel` or something like that to avoid conflicting with someone use of variables of the same name on window.

draffensperger commented on packages/opentelemetry-core/src/common/types.ts:
```diff
@@ -22,6 +22,12 @@ export enum LogLevel {
+ * This is equivalent to:
+ * type LogLevelString = 'ERROR' | 'WARN' | 'INFO' | 'DEBUG';
+ */
+export type LogLevelString = keyof typeof LogLevel;
```
Very cool you can do this!

naseemkullah replied:
Yeah, I was really happy to stumble upon it here  https://www.typescriptlang.org/docs/handbook/enums.html#enums-at-compile-time 😄

naseemkullah commented:
@draffensperger @obecny @dyladan 

I think I will be out of depth in implementing the web version of this, but am willing to provided my hand be held along the way! If not may I suggest addressing the web counterpart in another PR by someone more capable?

draffensperger commented:
What about creating a GitHub issue to track setting the log level for the web (that references the discussion here), and then also reference that issue as a todo comment in the code for the web part?

dyladan commented on packages/opentelemetry-core/src/platform/node/log-level.ts:
```diff
@@ -0,0 +1,26 @@
+import { LogLevel, LogLevelString } from '../../common/types';
+
+export function getLogLevel(): LogLevel {
+  return LogLevel[(process.env.OTEL_LOG_LEVEL as LogLevelString) ?? 'INFO'];
```
if an invalid string is provided, this will return undefined.
```suggestion
  return LogLevel[(process.env.OTEL_LOG_LEVEL as LogLevelString)] ?? LogLevel.INFO;
```

naseemkullah replied:
Ah yes good catch!

naseemkullah commented:
> What about creating a GitHub issue to track setting the log level for the web (that references the discussion here), and then also reference that issue as a todo comment in the code for the web part?

done and done!

@dyladan I understand we are waiting for spec but could we get this is as a working name for the env var until the name is finalized?

vmarchaud commented on packages/opentelemetry-core/src/platform/browser/log-level.ts:
```diff
@@ -0,0 +1,25 @@
+ * @returns the default log level for opentelemetry.
+ * TODO: https://github.com/open-telemetry/opentelemetry-js/issues/1096
+ */
+export function getLogLevel(): LogLevel {
```
I think it would make more sense to call this function `getDefaultLogLevel()`

naseemkullah replied:
Yeah, both these names are slightly misleading since it gets the default level or the (env/global*) var set by user log level.  With that said you still prefer   `getDefaultLogLevel()` as the least bad of the two?            

* global is TODO.. so effectively it is just returning the default log level today in the case of browser

vmarchaud replied:
I would prefer `getDefaultLogLevel` indeed. It implies that it can be overriden for each component if needed

dyladan commented on packages/opentelemetry-core/src/common/ConsoleLogger.ts:
```diff
@@ -16,9 +16,10 @@
 
 export class ConsoleLogger implements Logger {
-  constructor(level: LogLevel = LogLevel.INFO) {
+  constructor(level: LogLevel = getDefaultLogLevel()) {
```
I don't think default is a great name for this because if a user configures a log level it isn't the default. Maybe `getEnvironmentLogLevel` or something? You could even do something like this:
```suggestion
  constructor(level: LogLevel = getEnvironmentLogLevel() ?? LogLevel.INFO) {
```

naseemkullah replied:
that is starting to make more sense. getEnvLogLvl for browser would just return `undefined` today?

dyladan replied:
yes

naseemkullah replied:
nullish  coalescer broke tests, had to use or 🤷‍♂️

naseemkullah replied:
> I don't think default is a great name for this because if a user configures a log level it isn't the default. Maybe `getEnvironmentLogLevel` or something? You could even do something like this:

I am second guessing this approach now since we have to validate `getEnv...() ??/|| LogLevel.Info` in various places (core, tracing and metrics repos) now, whereas if we encompass the "env var set or default" in the platform function we only need to worry about its functionality in the core repo. Does that make sense?

naseemkullah replied:
As such: https://github.com/open-telemetry/opentelemetry-js/pull/974/commits/71627461ddc147b6031faafd0c715c7257f0864c

dyladan replied:
Null coalescing doesn't work in the browser tests. Probably has to do with the version of ts or configuration used by the webpacking. I'm fine with the new impl.

naseemkullah commented:
Going to squash commits as they are getting a little out of hand...

naseemkullah commented:
OK history has been cleaned up, this is ready for review and merging if we agree to keep `OTEL_LOG_LEVEL` as a working name until spec says otherwise.

naseemkullah commented:
Removed a test that should not have been ( an unwanted remanent of the history cleanup), removed it and squashed to keep clean history. Sorry, this should be it now.

iliasbhal commented on packages/opentelemetry-core/src/platform/browser/log-level.ts:
```diff
@@ -0,0 +1,25 @@
+ * TODO: https://github.com/open-telemetry/opentelemetry-js/issues/1096
+ */
+export function getEnvironmentalOrDefaultLogLevel(): LogLevel {
+  return LogLevel.INFO;
```
I'm going to work on detecting OTEL_LOG_LEVEL within the browser
( WIP, I'm going to create a suggestion )

dyladan replied:
I think actually it makes sense to merge this first. We may hold release, but you've already done the work here to split the browser/node impls which @iliasbhal will have to do if this isn't merged.

naseemkullah replied:
ok that works too!

iliasbhal commented on packages/opentelemetry-core/src/platform/browser/log-level.ts:
```diff
@@ -0,0 +1,25 @@
+ * TODO: https://github.com/open-telemetry/opentelemetry-js/issues/1096
+ */
+export function getEnvironmentalOrDefaultLogLevel(): LogLevel {
+  return LogLevel.INFO;
```
This should be compatible with bundlers like Webpack ( that inject a process variable into the bundle ) and still work with vanilla projects.

```suggestion
  const processLogLevel = (typeof process !== 'undefined' && process.env && process.env.OTEL_LOG_LEVEL)
    || window.OTEL_LOG_LEVEL;

    return LogLevel[processLogLevel as LogLevelString]
      ||  LogLevel.INFO
}

```

naseemkullah replied:
@obecny @draffensperger could you please review this suggestion by @iliasbhal ?

obecny replied:
I think we will end up with many code duplication. I added my suggestion here: https://github.com/open-telemetry/opentelemetry-js/pull/974#issuecomment-638420365

I also think that we should probably have enum to be used with environment variables. And if we have platform specific environment, then it will be a matter of reading the log level in one place only without worrying if this is a window, global, process.env or something else.

iliasbhal commented on packages/opentelemetry-core/src/platform/node/log-level.ts:
```diff
@@ -0,0 +1,33 @@
+ * if not, falls back to INFO.
+ */
+export function getEnvironmentalOrDefaultLogLevel(): LogLevel {
+  return process.env.OTEL_LOG_LEVEL &&
+    process.env.OTEL_LOG_LEVEL.toUpperCase().match('ERROR|WARN|INFO|DEBUG')
+    ? LogLevel[process.env.OTEL_LOG_LEVEL as LogLevelString]
+    : LogLevel.INFO;
```
No need to check manually here, it's going to fallback without a dev having to maintain this portion of the code.

```suggestion
  return LogLevel[process.env.OTEL_LOG_LEVEL as LogLevelString]
    || LogLevel.INFO;
```

obecny commented:
What if you do it the same way we have a `global` now. Create platform specific `environment` and then use it for checking of variables existence. So it will be not only log level, but we will already gain a support for other params that will be unified across all platforms. Having this you will add a first param for log level. It will be much easier to introduce any new env - just a new enum value. WDYT ?

iliasbhal commented on packages/opentelemetry-core/src/platform/node/log-level.ts:
```diff
@@ -0,0 +1,33 @@
+ */
+export function getEnvironmentalOrDefaultLogLevel(): LogLevel {
+  return process.env.OTEL_LOG_LEVEL &&
+    process.env.OTEL_LOG_LEVEL.toUpperCase().match('ERROR|WARN|INFO|DEBUG')
```
lowercased log-level string will return `undefined`.
```
console.log(upperCasedLogLevel) // ["WARN"]
console.log(process.env.OTEL_LOG_LEVEL); // 'warn'
console.log(LogLevel[process.env.OTEL_LOG_LEVEL]); // undefined because 'warn' is not a key in LogLevel
```

naseemkullah replied:
true

obecny commented:
@dyladan does this still need to wait for spec or not anymore ?

dyladan commented:
No. There was a spec issue tracking these things, but it doesn't seem to be going anywhere and other SIGs are just adding ENV configs.

dyladan commented:
@naseemkullah please fix the merge conflict

naseemkullah commented:
I've made a change: I've set it such that a LogLevelString is expected (upper or lower case) instead of a number (0,1,2,3). WDYT @obecny since you wrote the env parser?

obecny commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -37,7 +37,7 @@ const ENVIRONMENT_NUMBERS: Partial<keyof ENVIRONMENT>[] = [
 export const DEFAULT_ENVIRONMENT: Required<ENVIRONMENT> = {
   OTEL_NO_PATCH_MODULES: '',
-  OTEL_LOG_LEVEL: LogLevel.ERROR,
+  OTEL_LOG_LEVEL: LogLevel.INFO,
```
are you sure about this? - it might show too many data in console

naseemkullah replied:
It's true but this is based on #856 , has `ERROR` been agreed upon since as what should be the default level?

dyladan replied:
info is correct for now. I think we were going to change it to warn when we go GA

mayurkale22 replied:
I think we should open an issue to change the log level to warn and add `release:required-for-ga` label.

obecny commented on packages/opentelemetry-core/test/common/ConsoleLogger.test.ts:
```diff
@@ -114,5 +116,33 @@ describe('ConsoleLogger', () => {
       assert.strictEqual(debugCalledArgs, null);
     });
+
+    if (typeof process !== 'undefined' && process.release.name === 'node')
```
process is not available in browser so we shouldn't imho have test that is meant only for node that is run in browser too.

naseemkullah replied:
true, I am not sure how to set the env in the test for both node and browser, that is why I just test for node with `        process.env.OTEL_LOG_LEVEL = ...;`, how can I set an env var for both platforms in the test?

obecny replied:
can you try to mock/overwrite getEnv or parseEnvironment so it will be platform unindependent ?

obecny commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -64,6 +64,26 @@ function parseNumber(
+    const logLevelString = value.toUpperCase() as LogLevelString;
+    logLevel = LogLevel[logLevelString];
+  }
+  environment[key] = logLevel || DEFAULT_ENVIRONMENT[key];
```
the defaults are attached in getEnv()
here only the correct existing value should be attached

```
if (value && typeof value === 'string') {
    const logLevelString = value.toUpperCase() as LogLevelString;
    environment[key] = LogLevel[logLevelString];
  }
```

naseemkullah replied:
True, the reason I did this is to deal with invalid string values.
But we can do a more thorough check that upper cased `value` is one of the LogLevelStrings

naseemkullah replied:
updated!

obecny replied:
I was only referring to logic that default value is set in different place here you basically set the environment only if it passed all your different checks, validation etc. otherwise you don't set it and it will fallback to default automatically

naseemkullah replied:
yup that was a good point, the validation was made stricter and the extraneous fall back has been removed

dyladan commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -64,6 +64,28 @@ function parseNumber(
+ * @param environment
+ * @param values
+ */
+function parseLogLevel(
```
This function seems quite convoluted. Why not just have a switch/case?

dyladan replied:
Is this function modifying the environment? Why not a function that just returns a log level and use that?

naseemkullah replied:
switch/case sounds good, I was trying to be clever and get the values of logLevelStrings without repeating them (`DEBUG`, `INFO`, `WARN` and `ERROR`) but it's too much convolution as mentioned.

obecny replied:
`['DEBUG', 'INFO', ....].indexOf(checkMe)>=0` isn't enough ?

naseemkullah replied:
I went with the switch case for now, 

@obecny I do not fully understand this suggestion could you please explain?

obecny replied:
was just referring to validation with regexp etc., loglevel can have 4 values only

dyladan commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -64,6 +64,28 @@ function parseNumber(
+  const logLevelStrings = Object.keys(LogLevel).filter(
+    k => isNaN(Number(k)) === true
+  ) as LogLevelString[];
+  const regex = new RegExp(logLevelStrings.join('|'), 'i');
```
Dynamically built regexes are rarely a good idea in my experience.

naseemkullah replied:
any anecdote in particular to share?

dyladan replied:
Not in particular

dyladan commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -64,6 +64,28 @@ function parseNumber(
+) {
+  const value = values[key];
+  const logLevelStrings = Object.keys(LogLevel).filter(
+    k => isNaN(Number(k)) === true
```
What is this filter doing? looking for keys of the enum that aren't numbers?

naseemkullah replied:
right, I almost put a comment, guess I should have `Object.keys(LogLevel)` returns not only `INFO` and such but also `0` `1` `2` `3`

dyladan commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -79,7 +101,7 @@ export function parseEnvironment(values: ENVIRONMENT_MAP): ENVIRONMENT {
 
       case 'OTEL_LOG_LEVEL':
-        parseNumber(key, environment, values, LogLevel.ERROR, LogLevel.DEBUG);
+        parseLogLevel(key, environment, values);
```
Really don't love the side-effect driven nature of this.

naseemkullah replied:
I don't understand what is meant by that? Could you please elaborate?

dyladan replied:
You run a function which has no output, but modifies global state (the environment). This state then affects the behavior of the program down the line. Instead, I think a better solution is to have a function which looks at the env and returns a loglevel. This function is then simply called during the creation of a logger and the environment is never modified.

naseemkullah replied:
I'm starting to get it, the `parseNumber()` function used for OTEL_SAMPLING_PROBABILITY (and previously to this PR for OTEL_LOG_LEVEL), is a side-effect driver function or not though?

dyladan replied:
Yes. `parseDriver` modifies global state, which is the definition of side-effect.

naseemkullah replied:
you mean `parseNumber`

naseemkullah replied:
So, I based the log level setting function on the `parseNumber` behaviour which is side-effect driven.

Should I do something non side-effect driven for log level in this PR?

dyladan replied:
Yes. I am suggesting something like the following:

```typescript
function getLogLevelFromEnv() {
  // no need to pass key and values in since it will always come from the same var
  const value = environment.OTEL_LOG_LEVEL;
  switch (typeof value === 'string' ? value.toUpperCase() : value) {
    case 'DEBUG':
      return LogLevel.DEBUG;

    case 'INFO':
      return LogLevel.INFO;

    case 'WARN':
      return LogLevel.WARN;

    case 'ERROR':
      return LogLevel.ERROR;

    default:
      return null;
  }
}
```

This function to be used when determining which log level to use when creating a logger.

naseemkullah replied:
very nice, lastly, this new function should be exported, and we will not use `getEnv()` anymore for logger instantiation, but rather this new function?

obecny replied:
The whole idea of having the getEnv() is to have one place to get you all env parsed, validated with default values etc. If you use any of this function then you endup of problem for cross platform and repeating the same code again and again in many places. So using any of "internal" function just doesn't make sense with this whole idea.

obecny replied:
With regards to "side effect" we have exactly the same pattern with injecting / extracting context. This is correct here, too you just want to have one function that will validate the setting and if the setting is valid then add that to the env so that default value doesn't overwrite it later. Not sure why then "side effect" is wrong here ?

naseemkullah replied:
Please advise @dyladan, @obecny raises a good point

dyladan replied:
Imagine the following case:

1. user sets the env var to "warning" which is invalid
2. user sees info level logs
3. as part of debugging, they print the env var and see "2" because we have overwritten it, which is confusing
4. they spend time trying to figure out why their env var is failing to be set, when the real problem is that the value itself is wrong

If others think this isn't a problem, then I am willing to concede.

obecny replied:
You are mixing here 2 things, in this function (helper internal function) you are adding a new param to private internal variable not a process.env - nothing modifies that.

dyladan replied:
ah ok

obecny replied:
https://github.com/open-telemetry/opentelemetry-js/blob/e9b2cf9aeb1daf5ffbab800681bfe1cafc636576/packages/opentelemetry-api/src/context/propagation/setter.ts#L29

obecny replied:
and here
https://github.com/open-telemetry/opentelemetry-js/blob/e9b2cf9aeb1daf5ffbab800681bfe1cafc636576/packages/opentelemetry-core/src/context/propagation/B3Propagator.ts#L47

naseemkullah replied:
That's a good example @dyladan, @obecny are you holding your ground on this one? I'll let you make the final call (for now at least) if nobody else wants to weigh in.

obecny replied:
I have a feeling we are talking about 2 different thing, If I understand correctly, @dyladan was thinking that `parse...` helpers function modifies the `process.env` which is not the case at all. All the helpers are doing they are just adding / injecting values into a variable that you tell it to inject. In this example this is a new variable that takes all config/env variables from either process.env or window on browser and returns a valid values (validated, defaulted etc.) but in both cases it doesn't modify the process.env or window at all. In fact you can provide any obj and tell it to give you the validated params. The whole pattern how the helpers are behaving is the same approach what we use in propagators. You create a new carrier (our internal env variable) and tell it to inject all valid environment params.

naseemkullah replied:
ok so if I understand correctly, @dyladan was afraid that we were modified `process.env` or browser equivalent window, however we never were, and so there is no danger of `side-effect`? as such current implementation is ok?

dyladan replied:
yes. i was mistaken

naseemkullah replied:
In fact we are simply modifying an internal env var/carrier called `environment` right?

obecny replied:
I don't know lets @dyladan comment, I'm calling a day (friday night btw) for me today ;)

naseemkullah replied:
Have a great weekend @obecny 🕺 ! And thanks for the explanations.

iliasbhal commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -64,6 +64,28 @@ function parseNumber(
+  const logLevelStrings = Object.keys(LogLevel).filter(
+    k => isNaN(Number(k)) === true
+  ) as LogLevelString[];
+  const regex = new RegExp(logLevelStrings.join('|'), 'i');
+  if (value && typeof value === 'string' && regex.test(value)) {
```
@naseemkullah would that work here?

```suggestion
  const regex = new RegExp(logLevelStrings.join('|'), 'i');
  if (value && typeof value === 'string' && logLevelStrings.includes(value)) {
```

naseemkullah replied:
Probably @iliasbhal but I just changed the function to a simpler switch case 😅

naseemkullah commented:
Please excuse the force pushes, it is weekend and I presume nobody will be re-reviewing until Monday and would like a clean commit history by then.

obecny commented on packages/opentelemetry-core/src/utils/environment.ts:
```diff
@@ -95,3 +129,37 @@ export function parseEnvironment(values: ENVIRONMENT_MAP): ENVIRONMENT {
+/**
+ * Mocks environment used for tests.
+ */
+export function mockEnvironment(values: ENVIRONMENT_MAP) {
```
@dyladan ^^

naseemkullah replied:
So the idea was to use it in multiple test files. While moving it here and exporting it so that multiple test files could use it seems to work it might not be the best solution.

We had a test-utils package but that was moved to contrib repo I believe. Should I copy paste these functions in both test files?

obecny replied:
but it was inside file `/test/utils/environment.test.ts ` which is perfectly fine, by moving this to `environment.ts ` mean we are exporting a test function as our api which is not something we want

dyladan replied:
test-utils is in the contrib repo because it was only used by contrib plugins, but it might be a good idea to move it back into the main repo. I have said on other PRs I don't like the idea of exporting testing/fake/mock behavior from modules because users _will_ find them and they _will_ use them even if we dont want them to, and they will be confused and we will have to deal with it.

obecny replied:
so you can still easily share this function between tests in the same package, just don't export them as our sdk

naseemkullah replied:
Thanks I will change accordingly as not to export the function

naseemkullah replied:
Alright, exported from its original location (test file itself) not sure why i did not do that first time around, think I was fatigued at that time.

obecny replied:
just fix the conflicts pls
